File: validators/v3.0.0/performance_config_v3.0.0.ts
/**
 * Performance Configuration for SDUI Web Validator v3.0.0
 * TypeScript конфигурация производительности
 *
 * @version 3.0.0
 * @author Performance Engineer Agent
 * @date 2025-10-05
 */

export interface PerformanceTargets {
  /** Целевое время для файлов <100KB (мс) */
  smallFile: number;
  /** Целевое время для файлов 100KB-1MB (мс) */
  mediumFile: number;
  /** Целевое время для файлов >1MB (мс) */
  largeFile: number;
  /** Целевое время для batch валидации 100 файлов (мс) */
  batch100Files: number;
}

export interface CacheConfig {
  /** Включить кэширование схем */
  schemaCache: boolean;
  /** Максимальный размер кэша схем */
  schemaCacheMaxSize: number;
  /** Включить кэширование position map */
  positionMapCache: boolean;
  /** Максимальный размер кэша position map */
  positionMapCacheMaxSize: number;
  /** TTL (Time To Live) для кэша в секундах */
  cacheTTL: number;
}

export interface FileProcessingConfig {
  /** Включить ленивую загрузку */
  lazyLoading: boolean;
  /** Потоковая обработка больших файлов */
  streamLargeFiles: boolean;
  /** Порог размера файла для потоковой обработки (KB) */
  largeFileThresholdKB: number;
  /** Размер буфера чтения (KB) */
  readBufferSizeKB: number;
}

export interface ParallelConfig {
  /** Включить параллельную валидацию */
  enabled: boolean;
  /** Максимальное количество воркеров (null = CPU count) */
  maxWorkers: number | null;
  /** Размер батча для параллельной обработки */
  batchSize: number;
  /** Стратегия распределения задач */
  schedulingStrategy: 'round-robin' | 'work-stealing' | 'dynamic';
}

export interface MemoryConfig {
  /** Включить memory pooling */
  poolingEnabled: boolean;
  /** Максимальный размер памяти (MB) */
  maxMemoryMB: number;
  /** Порог для сборки мусора (количество файлов) */
  gcThreshold: number;
  /** Включить агрессивную сборку мусора */
  aggressiveGC: boolean;
}

export interface OptimizationFlags {
  /** Использовать скомпилированные регулярные выражения */
  useCompiledRegex: boolean;
  /** Использовать индекс строк */
  useLineIndex: boolean;
  /** Использовать индекс схем */
  useSchemaIndex: boolean;
  /** Пропускать валидацию при попадании в кэш */
  skipValidationOnCacheHit: boolean;
  /** Включить профилирование производительности */
  enableProfiling: boolean;
}

export interface MonitoringConfig {
  /** Включить мониторинг производительности */
  enabled: boolean;
  /** Порог предупреждения (мс) */
  warningThresholdMS: number;
  /** Порог ошибки (мс) */
  errorThresholdMS: number;
  /** Логировать медленные операции */
  logSlowOperations: boolean;
  /** Экспортировать метрики */
  exportMetrics: boolean;
  /** Формат экспорта метрик */
  metricsFormat: 'json' | 'prometheus' | 'csv';
}

export interface PerformanceConfig {
  /** Целевые метрики производительности */
  targets: PerformanceTargets;
  /** Настройки кэширования */
  cache: CacheConfig;
  /** Настройки обработки файлов */
  fileProcessing: FileProcessingConfig;
  /** Настройки параллелизма */
  parallel: ParallelConfig;
  /** Настройки управления памятью */
  memory: MemoryConfig;
  /** Флаги оптимизаций */
  optimizations: OptimizationFlags;
  /** Настройки мониторинга */
  monitoring: MonitoringConfig;
}

/**
 * Production конфигурация - максимальная производительность
 */
export const productionConfig: PerformanceConfig = {
  targets: {
    smallFile: 100,
    mediumFile: 500,
    largeFile: 2000,
    batch100Files: 10000,
  },
  cache: {
    schemaCache: true,
    schemaCacheMaxSize: 512,
    positionMapCache: true,
    positionMapCacheMaxSize: 256,
    cacheTTL: 3600,
  },
  fileProcessing: {
    lazyLoading: true,
    streamLargeFiles: true,
    largeFileThresholdKB: 500,
    readBufferSizeKB: 64,
  },
  parallel: {
    enabled: true,
    maxWorkers: null, // auto-detect CPU count
    batchSize: 10,
    schedulingStrategy: 'work-stealing',
  },
  memory: {
    poolingEnabled: true,
    maxMemoryMB: 1024,
    gcThreshold: 100,
    aggressiveGC: false,
  },
  optimizations: {
    useCompiledRegex: true,
    useLineIndex: true,
    useSchemaIndex: true,
    skipValidationOnCacheHit: false,
    enableProfiling: false,
  },
  monitoring: {
    enabled: true,
    warningThresholdMS: 100,
    errorThresholdMS: 500,
    logSlowOperations: true,
    exportMetrics: true,
    metricsFormat: 'json',
  },
};

/**
 * Development конфигурация - баланс производительности и отладки
 */
export const developmentConfig: PerformanceConfig = {
  targets: {
    smallFile: 100,
    mediumFile: 500,
    largeFile: 2000,
    batch100Files: 10000,
  },
  cache: {
    schemaCache: true,
    schemaCacheMaxSize: 128,
    positionMapCache: true,
    positionMapCacheMaxSize: 64,
    cacheTTL: 1800,
  },
  fileProcessing: {
    lazyLoading: true,
    streamLargeFiles: false,
    largeFileThresholdKB: 500,
    readBufferSizeKB: 32,
  },
  parallel: {
    enabled: false, // отключено для упрощения отладки
    maxWorkers: 2,
    batchSize: 5,
    schedulingStrategy: 'round-robin',
  },
  memory: {
    poolingEnabled: false,
    maxMemoryMB: 512,
    gcThreshold: 50,
    aggressiveGC: false,
  },
  optimizations: {
    useCompiledRegex: true,
    useLineIndex: true,
    useSchemaIndex: true,
    skipValidationOnCacheHit: false,
    enableProfiling: true,
  },
  monitoring: {
    enabled: true,
    warningThresholdMS: 50,
    errorThresholdMS: 200,
    logSlowOperations: true,
    exportMetrics: false,
    metricsFormat: 'json',
  },
};

/**
 * Minimal конфигурация - без оптимизаций (для тестирования)
 */
export const minimalConfig: PerformanceConfig = {
  targets: {
    smallFile: 100,
    mediumFile: 500,
    largeFile: 2000,
    batch100Files: 10000,
  },
  cache: {
    schemaCache: false,
    schemaCacheMaxSize: 0,
    positionMapCache: false,
    positionMapCacheMaxSize: 0,
    cacheTTL: 0,
  },
  fileProcessing: {
    lazyLoading: false,
    streamLargeFiles: false,
    largeFileThresholdKB: 0,
    readBufferSizeKB: 16,
  },
  parallel: {
    enabled: false,
    maxWorkers: 1,
    batchSize: 1,
    schedulingStrategy: 'round-robin',
  },
  memory: {
    poolingEnabled: false,
    maxMemoryMB: 256,
    gcThreshold: 10,
    aggressiveGC: false,
  },
  optimizations: {
    useCompiledRegex: false,
    useLineIndex: false,
    useSchemaIndex: false,
    skipValidationOnCacheHit: false,
    enableProfiling: false,
  },
  monitoring: {
    enabled: false,
    warningThresholdMS: 1000,
    errorThresholdMS: 5000,
    logSlowOperations: false,
    exportMetrics: false,
    metricsFormat: 'json',
  },
};

/**
 * CI/CD конфигурация - для автоматизированного тестирования
 */
export const ciConfig: PerformanceConfig = {
  targets: {
    smallFile: 100,
    mediumFile: 500,
    largeFile: 2000,
    batch100Files: 10000,
  },
  cache: {
    schemaCache: true,
    schemaCacheMaxSize: 256,
    positionMapCache: true,
    positionMapCacheMaxSize: 128,
    cacheTTL: 300,
  },
  fileProcessing: {
    lazyLoading: true,
    streamLargeFiles: true,
    largeFileThresholdKB: 500,
    readBufferSizeKB: 64,
  },
  parallel: {
    enabled: true,
    maxWorkers: 4, // фиксированное количество для стабильности
    batchSize: 10,
    schedulingStrategy: 'round-robin',
  },
  memory: {
    poolingEnabled: true,
    maxMemoryMB: 512,
    gcThreshold: 50,
    aggressiveGC: true,
  },
  optimizations: {
    useCompiledRegex: true,
    useLineIndex: true,
    useSchemaIndex: true,
    skipValidationOnCacheHit: false,
    enableProfiling: true,
  },
  monitoring: {
    enabled: true,
    warningThresholdMS: 100,
    errorThresholdMS: 500,
    logSlowOperations: true,
    exportMetrics: true,
    metricsFormat: 'json',
  },
};

/**
 * Benchmark статистика
 */
export interface BenchmarkStats {
  category: 'small' | 'medium' | 'large';
  count: number;
  mean: number;
  median: number;
  min: number;
  max: number;
  p95: number;
  p99: number;
}

/**
 * Результаты валидации метрик
 */
export interface MetricValidationResult {
  category: string;
  target: number;
  actual: number;
  passed: boolean;
  margin?: number;
}

/**
 * Проверяет соответствие фактических метрик целевым
 */
export function validateTargets(
  config: PerformanceConfig,
  actualStats: Record<string, BenchmarkStats>
): MetricValidationResult[] {
  const results: MetricValidationResult[] = [];

  const categoryTargets: Record<string, number> = {
    small: config.targets.smallFile,
    medium: config.targets.mediumFile,
    large: config.targets.largeFile,
  };

  for (const [category, targetMS] of Object.entries(categoryTargets)) {
    if (category in actualStats) {
      const actualMS = actualStats[category].median;
      const passed = actualMS <= targetMS;
      const margin = passed ? ((targetMS - actualMS) / targetMS) * 100 : undefined;

      results.push({
        category,
        target: targetMS,
        actual: actualMS,
        passed,
        margin,
      });
    }
  }

  return results;
}

/**
 * Форматирует результаты валидации в читаемый вид
 */
export function formatValidationResults(results: MetricValidationResult[]): string {
  const lines: string[] = [
    '## Target Metrics Validation',
    '',
    '| Category | Target | Actual | Status | Margin |',
    '|----------|--------|--------|--------|--------|',
  ];

  for (const result of results) {
    const status = result.passed ? '✅ PASS' : '❌ FAIL';
    const margin = result.margin !== undefined ? `${result.margin.toFixed(1)}%` : 'N/A';

    lines.push(
      `| ${result.category.padEnd(8)} | ${result.target.toString().padStart(6)}ms | ${result.actual
        .toFixed(2)
        .padStart(8)}ms | ${status} | ${margin.padStart(6)} |`
    );
  }

  return lines.join('\n');
}

/**
 * Экспортирует конфигурацию в формат для Python валидатора
 */
export function exportToPython(config: PerformanceConfig): string {
  return `# Auto-generated from performance_config_v3.0.0.ts
# Do not edit manually

PERFORMANCE_CONFIG = {
    'targets': {
        'small_file': ${config.targets.smallFile},
        'medium_file': ${config.targets.mediumFile},
        'large_file': ${config.targets.largeFile},
        'batch_100_files': ${config.targets.batch100Files},
    },
    'cache': {
        'schema_cache': ${config.cache.schemaCache},
        'schema_cache_max_size': ${config.cache.schemaCacheMaxSize},
        'position_map_cache': ${config.cache.positionMapCache},
        'position_map_cache_max_size': ${config.cache.positionMapCacheMaxSize},
        'cache_ttl': ${config.cache.cacheTTL},
    },
    'file_processing': {
        'lazy_loading': ${config.fileProcessing.lazyLoading},
        'stream_large_files': ${config.fileProcessing.streamLargeFiles},
        'large_file_threshold_kb': ${config.fileProcessing.largeFileThresholdKB},
        'read_buffer_size_kb': ${config.fileProcessing.readBufferSizeKB},
    },
    'parallel': {
        'enabled': ${config.parallel.enabled},
        'max_workers': ${config.parallel.maxWorkers === null ? 'None' : config.parallel.maxWorkers},
        'batch_size': ${config.parallel.batchSize},
        'scheduling_strategy': '${config.parallel.schedulingStrategy}',
    },
    'memory': {
        'pooling_enabled': ${config.memory.poolingEnabled},
        'max_memory_mb': ${config.memory.maxMemoryMB},
        'gc_threshold': ${config.memory.gcThreshold},
        'aggressive_gc': ${config.memory.aggressiveGC},
    },
    'optimizations': {
        'use_compiled_regex': ${config.optimizations.useCompiledRegex},
        'use_line_index': ${config.optimizations.useLineIndex},
        'use_schema_index': ${config.optimizations.useSchemaIndex},
        'skip_validation_on_cache_hit': ${config.optimizations.skipValidationOnCacheHit},
        'enable_profiling': ${config.optimizations.enableProfiling},
    },
    'monitoring': {
        'enabled': ${config.monitoring.enabled},
        'warning_threshold_ms': ${config.monitoring.warningThresholdMS},
        'error_threshold_ms': ${config.monitoring.errorThresholdMS},
        'log_slow_operations': ${config.monitoring.logSlowOperations},
        'export_metrics': ${config.monitoring.exportMetrics},
        'metrics_format': '${config.monitoring.metricsFormat}',
    },
}
`;
}

/**
 * Пример использования
 */
export function example() {
  // Production режим
  console.log('Production Config:');
  console.log(`  Schema Cache: ${productionConfig.cache.schemaCacheMaxSize}`);
  console.log(`  Parallel: ${productionConfig.parallel.enabled}`);
  console.log(`  Memory Pool: ${productionConfig.memory.poolingEnabled}`);

  // Валидация метрик
  const actualStats: Record<string, BenchmarkStats> = {
    small: {
      category: 'small',
      count: 10,
      mean: 0.62,
      median: 0.52,
      min: 0.04,
      max: 1.67,
      p95: 1.67,
      p99: 1.67,
    },
    medium: {
      category: 'medium',
      count: 5,
      mean: 2.63,
      median: 3.32,
      min: 1.51,
      max: 3.41,
      p95: 3.41,
      p99: 3.41,
    },
  };

  const validation = validateTargets(productionConfig, actualStats);
  console.log('\n' + formatValidationResults(validation));

  // Экспорт в Python
  console.log('\nPython export:');
  console.log(exportToPython(developmentConfig));
}

// Default export
export default {
  production: productionConfig,
  development: developmentConfig,
  minimal: minimalConfig,
  ci: ciConfig,
  validateTargets,
  formatValidationResults,
  exportToPython,
};

File: validators/v3.0.0/performance_config_v3.0.0.py
#!/usr/bin/env python3
"""
Performance Configuration for SDUI Web Validator v3.0.0
Конфигурация производительности с оптимизациями
"""

from dataclasses import dataclass
from typing import Optional
from functools import lru_cache
import os

@dataclass
class PerformanceConfig:
    """Конфигурация производительности валидатора"""

    # Cache settings
    schema_cache_enabled: bool = True
    schema_cache_max_size: int = 256
    position_map_cache_enabled: bool = True
    position_map_cache_max_size: int = 128

    # File processing
    lazy_loading_enabled: bool = True
    stream_large_files: bool = True
    large_file_threshold_kb: int = 500

    # Parallel processing
    parallel_validation_enabled: bool = True
    max_workers: Optional[int] = None  # None = CPU count
    batch_size: int = 10

    # Memory management
    memory_pooling_enabled: bool = True
    max_memory_mb: int = 512
    gc_threshold: int = 100  # файлов до сборки мусора

    # Performance targets (ms)
    target_small_file: int = 100      # <100KB
    target_medium_file: int = 500     # 100KB-1MB
    target_large_file: int = 2000     # >1MB
    target_batch_100: int = 10000     # 100 files

    # Optimization flags
    use_compiled_regex: bool = True
    use_line_index: bool = True
    skip_validation_on_cache_hit: bool = False

    @classmethod
    def from_env(cls) -> 'PerformanceConfig':
        """Создает конфигурацию из переменных окружения"""
        return cls(
            schema_cache_enabled=os.getenv('VALIDATOR_SCHEMA_CACHE', 'true').lower() == 'true',
            parallel_validation_enabled=os.getenv('VALIDATOR_PARALLEL', 'true').lower() == 'true',
            max_workers=int(os.getenv('VALIDATOR_MAX_WORKERS', '0')) or None,
            lazy_loading_enabled=os.getenv('VALIDATOR_LAZY_LOADING', 'true').lower() == 'true',
        )

    @classmethod
    def production(cls) -> 'PerformanceConfig':
        """Конфигурация для production среды"""
        return cls(
            schema_cache_max_size=512,
            position_map_cache_max_size=256,
            parallel_validation_enabled=True,
            memory_pooling_enabled=True,
        )

    @classmethod
    def development(cls) -> 'PerformanceConfig':
        """Конфигурация для development среды"""
        return cls(
            schema_cache_max_size=128,
            position_map_cache_max_size=64,
            parallel_validation_enabled=False,
            skip_validation_on_cache_hit=False,
        )

    @classmethod
    def minimal(cls) -> 'PerformanceConfig':
        """Минимальная конфигурация (без оптимизаций)"""
        return cls(
            schema_cache_enabled=False,
            position_map_cache_enabled=False,
            lazy_loading_enabled=False,
            parallel_validation_enabled=False,
            memory_pooling_enabled=False,
        )

    def validate_targets(self, actual: dict) -> dict:
        """Проверяет соответствие фактических метрик целевым"""
        results = {}

        targets = {
            'small': self.target_small_file,
            'medium': self.target_medium_file,
            'large': self.target_large_file,
        }

        for category, target_ms in targets.items():
            if category in actual:
                actual_ms = actual[category]['median']
                passed = actual_ms <= target_ms
                results[category] = {
                    'target': target_ms,
                    'actual': actual_ms,
                    'passed': passed,
                    'margin': ((target_ms - actual_ms) / target_ms) * 100 if passed else None
                }

        return results


# Singleton instance
_config: Optional[PerformanceConfig] = None

def get_config() -> PerformanceConfig:
    """Получает глобальную конфигурацию"""
    global _config
    if _config is None:
        _config = PerformanceConfig.from_env()
    return _config

def set_config(config: PerformanceConfig):
    """Устанавливает глобальную конфигурацию"""
    global _config
    _config = config

def reset_config():
    """Сбрасывает конфигурацию на дефолтную"""
    global _config
    _config = None


# Performance monitoring decorators
import time
from functools import wraps
from typing import Callable, Any

def profile_performance(threshold_ms: int = 100):
    """
    Декоратор для профилирования производительности функций

    Args:
        threshold_ms: Порог в миллисекундах для вывода предупреждения
    """
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs) -> Any:
            start = time.perf_counter()
            result = func(*args, **kwargs)
            end = time.perf_counter()

            duration_ms = (end - start) * 1000

            if duration_ms > threshold_ms:
                print(f"⚠️ Performance warning: {func.__name__} took {duration_ms:.2f}ms (threshold: {threshold_ms}ms)")

            return result
        return wrapper
    return decorator


def memoize_with_ttl(maxsize: int = 128, ttl_seconds: int = 3600):
    """
    Декоратор для мемоизации с TTL (Time To Live)

    Args:
        maxsize: Максимальный размер кэша
        ttl_seconds: Время жизни записи в кэше
    """
    def decorator(func: Callable) -> Callable:
        cache = {}
        cache_times = {}

        @wraps(func)
        def wrapper(*args, **kwargs):
            # Создаем ключ кэша
            key = (args, tuple(sorted(kwargs.items())))

            # Проверяем TTL
            if key in cache:
                age = time.time() - cache_times[key]
                if age < ttl_seconds:
                    return cache[key]
                else:
                    # Запись устарела, удаляем
                    del cache[key]
                    del cache_times[key]

            # Вычисляем результат
            result = func(*args, **kwargs)

            # Сохраняем в кэш (с учетом maxsize)
            if len(cache) >= maxsize:
                # Удаляем самую старую запись
                oldest_key = min(cache_times, key=cache_times.get)
                del cache[oldest_key]
                del cache_times[oldest_key]

            cache[key] = result
            cache_times[key] = time.time()

            return result

        # Добавляем метод для очистки кэша
        wrapper.cache_clear = lambda: (cache.clear(), cache_times.clear())
        wrapper.cache_info = lambda: {
            'size': len(cache),
            'maxsize': maxsize,
            'ttl': ttl_seconds
        }

        return wrapper
    return decorator


# Memory management utilities
import gc
import sys

class MemoryPool:
    """Пул памяти для больших объектов"""

    def __init__(self, max_size_mb: int = 512):
        self.max_size_mb = max_size_mb
        self.current_size = 0
        self.objects = []

    def allocate(self, obj: Any) -> bool:
        """Выделяет память для объекта"""
        obj_size = sys.getsizeof(obj) / 1024 / 1024  # MB

        if self.current_size + obj_size > self.max_size_mb:
            self.gc_collect()

        if self.current_size + obj_size <= self.max_size_mb:
            self.objects.append(obj)
            self.current_size += obj_size
            return True

        return False

    def gc_collect(self):
        """Принудительная сборка мусора"""
        self.objects.clear()
        gc.collect()
        self.current_size = 0

    def get_stats(self) -> dict:
        """Статистика использования памяти"""
        return {
            'current_mb': self.current_size,
            'max_mb': self.max_size_mb,
            'utilization': (self.current_size / self.max_size_mb) * 100,
            'objects_count': len(self.objects)
        }


# Example usage
if __name__ == '__main__':
    # Production config
    prod_config = PerformanceConfig.production()
    print("Production Config:")
    print(f"  Schema Cache: {prod_config.schema_cache_max_size}")
    print(f"  Parallel: {prod_config.parallel_validation_enabled}")
    print(f"  Memory Pool: {prod_config.memory_pooling_enabled}")

    # Validate against targets
    actual_stats = {
        'small': {'median': 0.52},
        'medium': {'median': 3.32},
    }

    validation = prod_config.validate_targets(actual_stats)
    print("\nTarget Validation:")
    for category, result in validation.items():
        status = "✅ PASS" if result['passed'] else "❌ FAIL"
        print(f"  {category}: {status} ({result['actual']:.2f}ms / {result['target']}ms)")
        if result['margin']:
            print(f"    Margin: {result['margin']:.1f}% under target")
